### PR TITLE
Add C# bindings for statistical charts, improve optional argument APIs

### PR DESCRIPTION
**Pull request by kMutagene**
This PR adds C# bindings for the following charts:

- [x] Histogram
- [x] Histogram2D
- [x] BoxPlot
- [x] Violin
- [x] Histogram2DContour
- [x] PointDensity
- [x] DensityMapbox
- [x] ContourCarpet

It also improves the optional argument usage to allow both value and reference types as generics where this is possible in the core API

Some GenericChart extension emthods were also added.

In general, the chart support matrix now looks like this:

(https://github.com/plotly/Plotly.NET/blob/329c50dbb8a8716800978b4e9f54a12286897ee6/tests/Plotly.NET.Tests.CSharpConsole/Program.cs)

image

@WhiteBlackGoose

### CODE REVIEW

File: src/Plotly.NET.CSharp/ChartAPI/Chart2D.cs [hunk 1]
```diff
@@ -436,5 +437,539 @@ public static partial class Chart
...
+        /// <param name="MarkerColor">Sets the color of the histogram's bars.</param>
+        /// <param name="Marker">Sets the marker for the histogram's bars (use this for more finegrained control than the other marker-associated arguments).</param>
+        /// <param name="Line">Sets the outline of the histogram's bars.</param>
+        /// <param name="XError">Sets the x error of this trace.</param>
+        /// <param name="YError">Sets the y error of this trace.</param>
+        /// <param name="Cumulative">Sets wether and how the cumulative distribution is displayed</param>
+        /// <param name="HoverLabel">Sets the style of the hoverlabels of this trace.</param>
+        /// <param name="UseDefaults">If set to false, ignore the global default settings set in `Defaults`</param>
+        public static GenericChart.GenericChart Histogram<XType, YType, TextType>(
+            [Optional] IEnumerable<XType>? X,
```
Comment (WhiteBlackGoose) [inline src/Plotly.NET.CSharp/ChartAPI/Chart2D.cs]: Why is this optional?

File: src/Plotly.NET.CSharp/GenericChartExtensions.cs [hunk 2]
```diff
@@ -62,8 +76,8 @@ public static class GenericChartExtensions
         /// Sets the size of a Chart (in pixels)
         public static GenericChart.GenericChart WithSize(
             this GenericChart.GenericChart gChart,
-            int? Width = null,
-            int? Height = null
+            int? Width,
+            int? Height
```
Comment (WhiteBlackGoose) [inline src/Plotly.NET.CSharp/GenericChartExtensions.cs]: Why removed null?

File: src/Plotly.NET.CSharp/GenericChartExtensions.cs [hunk 3]
```diff
@@ -103,39 +118,39 @@ public static class GenericChartExtensions
...
+            StyleParam.LinearAxisId? Overlaying,
+            Tuple<double, double>? Domain,
+            double? Position,
+            StyleParam.CategoryOrder? CategoryOrder,
+            IEnumerable<CategoryArrayType>? CategoryArray,
+            RangeSlider? RangeSlider,
+            RangeSelector? RangeSelector,
+            Color? BackgroundColor,
+            bool? ShowBackground,
+            StyleParam.SubPlotId? Id
```
Comment (WhiteBlackGoose) [inline src/Plotly.NET.CSharp/GenericChartExtensions.cs]: Shouldn't those be `[Optional]`?

File: src/Plotly.NET.CSharp/GenericChartExtensions.cs [hunk 4]
```diff
@@ -206,39 +222,39 @@ public static class GenericChartExtensions
...
```
Comment (WhiteBlackGoose) [inline src/Plotly.NET.CSharp/GenericChartExtensions.cs]: here too

File: src/Plotly.NET.CSharp/Helpers.cs [hunk 5]
```diff
@@ -15,7 +15,16 @@ static class Helpers {
...
+        static internal Microsoft.FSharp.Core.FSharpOption<T> ToOption<T>(this T? thing) {
+            if (EqualityComparer<T>.Default.Equals(thing, default(T)))
+            {
+                return FSharpOption<T>.None;
+            }
+            else
+            {
+                return new(thing);
+            }
+        }
```
Comment (WhiteBlackGoose) [inline src/Plotly.NET.CSharp/Helpers.cs]: Why this change? `T?` is not nullable if `T` is a value type

**Review by WhiteBlackGoose — COMMENTED** [review]
Also, by C# convention parameter names should be camel case. I don't think it's important, but in case you didn't know this

File: src/Plotly.NET.CSharp/GenericChartExtensions.cs (hunk 3 above)
Comment (kMutagene) [reply src/Plotly.NET.CSharp/GenericChartExtensions.cs]: oh absolutely

File: src/Plotly.NET.CSharp/ChartAPI/Chart2D.cs (hunk 1 above)
Comment (kMutagene) [reply src/Plotly.NET.CSharp/ChartAPI/Chart2D.cs]: you can visualize histograms across either x or y dimension, this is implemented the same in the core API

**Comment by kMutagene** [comment]
>Also, by C# convention parameter names should be camel case. I don't think it's important, but in case you didn't know this

My initial rationale here was to name the arguments exacly the same as in the core API, might change this later though

File: src/Plotly.NET.CSharp/Helpers.cs (hunk 5 above)
Comment (kMutagene) [reply src/Plotly.NET.CSharp/Helpers.cs]: i dont remember the exact repro for this, but iirc without this default check i have gotten default values for non-nullable generic types instead of `None`

Comment (WhiteBlackGoose) [reply src/Plotly.NET.CSharp/Helpers.cs]: Uhhhh I see the problem. But if you do it this way, a valid 0 will still be interpreted as default

Comment (kMutagene) [reply src/Plotly.NET.CSharp/Helpers.cs]: Well if i revert this, this line gets me the following wrongly set defaults instead of `None`:

```csharp
Chart.Column<int,string,string>(
    values: new int [] { 3,4 },
    Keys: new string [] { "first", "second"}
)
```

```js
"base":0,"width":0,
```

which leads to an empty graph (width = 0). This is a problem only when a single value is set, collections do not have this default value problem. So the defaults that get omitted are `0`, `0.0`, `""`, and `false`, and that only when we are talking about generics. I would argue that this is better than charts that are invisible per default.

**Comment by JakeRadMSFT** [comment]
Awesome! I was just coming here to check the status of Histogram!

**Comment by kMutagene** [comment]
Waiting on some feedback on #322 before i proceed to adapt the API surface in this PR

**Comment by kMutagene** [comment]
Ill leave #322 up for now, but decided to just use the custom optional type everywhere for now.
